src/Libraries/Microsoft.Extensions.AI.Abstractions/Functions/AIFunctionArguments.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+#pragma warning disable SA1111 // Closing parenthesis should be on line of last parameter
+#pragma warning disable SA1112 // Closing parenthesis should be on line of opening parenthesis
+#pragma warning disable SA1114 // Parameter list should follow declaration
+#pragma warning disable CA1710 // Identifiers should have correct suffix
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Represents arguments to be used with <see cref="AIFunction.InvokeAsync"/>.</summary>
+/// <remarks>
+/// <see cref="AIFunctionArguments"/> is a dictionary of name/value pairs that are used
+/// as inputs to an <see cref="AIFunction"/>. However, an instance carries additional non-nominal
+/// information, such as an optional <see cref="IServiceProvider"/> that can be used by
+/// an <see cref="AIFunction"/> if it needs to resolve any services from a dependency injection
+/// container.
+/// </remarks>
+public sealed class AIFunctionArguments : IDictionary<string, object?>, IReadOnlyDictionary<string, object?>
+{
+    /// <summary>The nominal arguments.</summary>
+    private readonly Dictionary<string, object?> _arguments;
+
+    /// <summary>Initializes a new instance of the <see cref="AIFunctionArguments"/> class.</summary>
+    public AIFunctionArguments()
+    {
+        _arguments = [];
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AIFunctionArguments"/> class containing
+    /// the specified <paramref name="arguments"/>.
+    /// </summary>
+    /// <param name="arguments">The arguments represented by this instance.</param>
+    /// <remarks>
+    /// The <paramref name="arguments"/> reference will be stored if the instance is
+    /// already a <see cref="Dictionary{TKey, TValue}"/>, in which case all dictionary
+    /// operations on this instance will be routed directly to that instance. If <paramref name="arguments"/>
+    /// is not a dictionary, a shallow clone of its data will be used to populate this
+    /// instance. A <see langword="null"/> <paramref name="arguments"/> is treated as an
+    /// empty dictionary.
+    /// </remarks>
+    public AIFunctionArguments(IDictionary<string, object?>? arguments)
+    {
+        _arguments =
+            arguments is null ? [] :
+            arguments as Dictionary<string, object?> ??
+            new Dictionary<string, object?>(arguments);
+    }
+
+    /// <summary>Gets or sets services optionally associated with these arguments.</summary>
+    public IServiceProvider? Services { get; set; }
```

Comment:
I take it abstractions now has taken a dependency on DI?

Reply:
No. IServiceProvider is in System.ComponentModel.